python/ray/serve/_private/serve_request_response.py

```diff
@@ -0,0 +1,152 @@
+from abc import ABC
+import grpc
+import logging
+import pickle
+
+from starlette.types import Receive, Scope, Send
+from typing import Any, Callable, List, Generator, Optional, Tuple
+
+from ray.actor import ActorHandle
+from ray.serve._private.constants import SERVE_LOGGER_NAME
+from ray.serve._private.utils import DEFAULT
+from ray.serve._private.common import gRPCRequest, StreamingHTTPRequest
+
+logger = logging.getLogger(SERVE_LOGGER_NAME)
+
+
+class ServeRequest(ABC):
+    """Base ServeRequest class to use in the common interface among proxies"""
+
+    pass
+
+
+class ASGIServeRequest(ServeRequest):
+    """ServeRequest implementation to wrap ASGI scope, receive, and send."""
+
+    def __init__(self, scope: Scope, receive: Receive, send: Send):
+        self.scope = scope
+        self.receive = receive
+        self.send = send
+
+    @property
+    def request_type(self) -> str:
+        return self.scope.get("type", "")
+
+    @property
+    def client(self) -> str:
+        return self.scope.get("client", "")
+
+    @property
+    def method(self) -> str:
+        return self.scope.get("method", "websocket").upper()
+
+    @property
+    def root_path(self) -> str:
+        return self.scope.get("root_path", "")
+
+    @property
+    def route_path(self) -> str:
+        return self.scope.get("path", "")[len(self.root_path) :]
+
+    @property
+    def path(self) -> str:
+        return self.scope.get("path", "")
+
+    @property
+    def headers(self) -> List[Tuple[bytes, bytes]]:
+        return self.scope.get("headers", [])
+
+    def set_path(self, path: str):
+        self.scope["path"] = path
+
+    def set_root_path(self, root_path: str):
+        self.scope["root_path"] = root_path
+
+    def request_object(self, proxy_handle) -> StreamingHTTPRequest:
+        return StreamingHTTPRequest(
+            pickled_asgi_scope=pickle.dumps(self.scope),
+            http_proxy_handle=proxy_handle,
+        )
+
+
+class gRPCServeRequest(ServeRequest):
+    """ServeRequest implementation to wrap gRPC request protobuf and metadata."""
+
+    def __init__(
+        self,
+        request_proto: Any,
+        context: "grpc._cython.cygrpc._ServicerContext",
+        match_target: Callable[[str], Optional[str]],
+        service_method: str,
+        stream: bool,
+    ):
+        self.request = request_proto
+        self.context = context
+        self.match_target = match_target
+        self.service_method = service_method
+        self.stream = stream
+        self.app_name = ""
+        self.route_path = ""
+        self.request_id = None
+        self.method_name = "__call__"
```

Comment:
is `__call__` being used for /-/routes and /-/healthz? looks like it is always overwritten when dealing with user service method.